cpp/src/arrow/python/builtin_convert.cc

```diff
@@ -756,211 +1078,227 @@ class StructConverter : public TypedConverterVisitor<StructBuilder, StructConver
 
   // Handle NA / NullType case
   if (real_type->id() == Type::NA) {
-    out->reset(new NullArray(size));
+    ArrayVector chunks = {std::make_shared<NullArray>(size)};
+    *out = std::make_shared<ChunkedArray>(chunks);
     return Status::OK();
   }
 
-  // Give the sequence converter an array builder
-  std::unique_ptr<ArrayBuilder> builder;
-  RETURN_NOT_OK(MakeBuilder(pool, real_type, &builder));
-  RETURN_NOT_OK(AppendPySequence(seq, size, real_type, builder.get(), from_pandas));
-  return builder->Finish(out);
-}
+  // Create the sequence converter, initialize with the builder
+  std::unique_ptr<SeqConverter> converter;
+  RETURN_NOT_OK(
+      GetConverter(real_type, options.from_pandas, strict_conversions, &converter));
+
+  // Create ArrayBuilder for type, then pass into the SeqConverter
+  // instance. The reason this is created here rather than in GetConverter is
+  // because of nested types (child SeqConverter objects need the child
+  // builders created by MakeBuilder)
+  std::unique_ptr<ArrayBuilder> type_builder;
+  RETURN_NOT_OK(MakeBuilder(options.pool, real_type, &type_builder));
+  RETURN_NOT_OK(converter->Init(type_builder.get()));
+
+  // Convert values
+  if (mask != nullptr && mask != Py_None) {
+    RETURN_NOT_OK(converter->AppendMultipleMasked(seq, mask, size));
+  } else {
+    RETURN_NOT_OK(converter->AppendMultiple(seq, size));
+  }
 
-Status ConvertPySequence(PyObject* obj, MemoryPool* pool, bool from_pandas,
-                         std::shared_ptr<Array>* out) {
-  return ConvertPySequenceReal(obj, -1, nullptr, pool, from_pandas, out);
-}
+  // Retrieve result. Conversion may yield one or more array values
+  std::vector<std::shared_ptr<Array>> chunks;
+  RETURN_NOT_OK(converter->GetResult(&chunks));
+
+  // TODO(wesm): ARROW-2814 comment. Is further casting ever necessary?
 
-Status ConvertPySequence(PyObject* obj, const std::shared_ptr<DataType>& type,
-                         MemoryPool* pool, bool from_pandas,
-                         std::shared_ptr<Array>* out) {
-  return ConvertPySequenceReal(obj, -1, &type, pool, from_pandas, out);
+  *out = std::make_shared<ChunkedArray>(chunks);
+  return Status::OK();
 }
 
-Status ConvertPySequence(PyObject* obj, int64_t size, MemoryPool* pool, bool from_pandas,
-                         std::shared_ptr<Array>* out) {
-  return ConvertPySequenceReal(obj, size, nullptr, pool, from_pandas, out);
+Status ConvertPySequence(PyObject* obj, const PyConversionOptions& options,
+                         std::shared_ptr<ChunkedArray>* out) {
+  return ConvertPySequence(obj, nullptr, options, out);
 }
 
-Status ConvertPySequence(PyObject* obj, int64_t size,
-                         const std::shared_ptr<DataType>& type, MemoryPool* pool,
-                         bool from_pandas, std::shared_ptr<Array>* out) {
-  return ConvertPySequenceReal(obj, size, &type, pool, from_pandas, out);
+Status ConvertPySequence(PyObject* obj, std::shared_ptr<Array>* out) {
```

Comment:
This variant is only used in C++ unit tests. Should we keep it?

Reply:
removing